java/flight/src/main/java/org/apache/arrow/flight/FlightStream.java

```diff
@@ -95,10 +95,30 @@ public Schema getSchema() {
     return schema;
   }
 
+  /**
+   * Get the provider for dictionaries in this stream.
+   *
+   * <p>Does NOT retain a reference to the underlying dictionaries. Dictionaries may be updated as the stream is read.
+   *
+   * @see #retainDictionaries()
+   */
   public DictionaryProvider getDictionaryProvider() {
     return dictionaries;
   }
 
+  /**
+   * Retain a reference to each of the dictionaries in the stream. Should be called after finishing reading the stream,
+   * but before closing.
+   *
+   * @return The current dictionaries in the stream.
+   */
+  public DictionaryProvider retainDictionaries() {
```

Comment:
Not the happiest with this API or implementation; open to suggestions.

Reply:
We have to swap out the entire provider, rather than just retaining a reference to the underlying buffers, as closing a FieldVector also clears it.

Reply:
I'm just ramping up on the Java dictionary stuff so I apologize if these are silly questions: Would it make sense to have dictionaryprovider have some mode where it handles retention? and maybe rename this to transferDictionaries().  This actually looks like it does the opposite of retention (it looks like it removes a reference?

Reply:
Yes, this API is rather confusing :/ there is a corresponding method that gets a reference to the provider, and this method removes this object's reference without decrementing the ref count, so that when this object is closed, it won't decrement the ref count or do any other cleanup. Unfortunately, we can't just have this method increment the ref count for you, because the close() implementation for dictionaries also clears the underlying buffers.

The DictionaryProvider API is rather sparse: there was more than one time where I wanted to be able to iterate all the dictionaries, but you can't do that directly. I think there's been long-running discussion for improving the Java dictionary APIs, maybe this is another point to clear up.

Reply:
> Yes, this API is rather confusing :/ there is a corresponding method that gets a reference to the provider, and this method removes this object's reference without decrementing the ref count, so that when this object is closed, it won't decrement the ref count or do any other cleanup. Unfortunately, we can't just have this method increment the ref count for you, because the close() implementation for dictionaries also clears the underlying buffers.

Is the other API used, can it be removed?  

I think the javadoc needs more details that the dictionaryprovider closed must be closed by the client and this method can only be called once in a meaningful way (maybe we shoujd enforce this).  Somehow a name like 'claimDictionaryOwnership' makes more sense to me might be too verbose (I think there is a better name our there).  retainDictionaries makes me think they are being retained on this object.

> The DictionaryProvider API is rather sparse: there was more than one time where I wanted to be able to iterate all the dictionaries, but you can't do that directly. I think there's been long-running discussion for improving the Java dictionary APIs, maybe this is another point to clear up.

Yes it would seem so.

Reply:
> Is the other API used, can it be removed?

We could remove it, or at least mark it as deprecated. (It was in 0.14.0, but all of this is fairly new.)

I will update the javadoc. Maybe `takeDictionaries`? (Rust semantics would be nice here...)

Reply:
I suppose the problem is - do we ever expect dictionary batches to be interleaved with record batches? If so, then the two methods are not redundant (the non-owning one is useful for processing streaming data, the owning one is useful for batch processing). As implemented, I don't think we can interleave messages like this, but it's possible in the future with delta dictionaries, right?

Reply:
we do expect interleaved in the future.